// @sitecore-jss/sitecore-jss-dev-tools/dist/cjs/manifest/generator/manifest.babel-shim.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.addDictionary = exports.addContent = exports.addRoute = exports.addRouteType = exports.addPlaceholder = exports.addTemplate = exports.addComponent = void 0;
// this is a shim to allow non-TypeScript consumers to take advantage of typings data when they cannot specify the
// type of the manifest var in a manifest source file, e.g.
// instead of export default (manifest: Manifest) => { ... }, you can use
// import { addComponent } from 'thislib';
// export default (manifest) => { addComponent(manifest, { data }); }
// DEPRECATED: favor using typing JSDoc instead, e.g. @param {Manifest} manifest Manifest instance to add components to
/**
 * @param {Manifest} manifest
 * @param {...any} components
 */
function addComponent(manifest, ...components) {
    manifest.addComponent(...components);
}
exports.addComponent = addComponent;
/**
 * @param {Manifest} manifest
 * @param {...any} templates
 */
function addTemplate(manifest, ...templates) {
    manifest.addTemplate(...templates);
}
exports.addTemplate = addTemplate;
/**
 * @param {Manifest} manifest
 * @param {...any} placeholders
 */
function addPlaceholder(manifest, ...placeholders) {
    manifest.addPlaceholder(...placeholders);
}
exports.addPlaceholder = addPlaceholder;
/**
 * @param {Manifest} manifest
 * @param {...any} routeTypes
 */
function addRouteType(manifest, ...routeTypes) {
    manifest.addRouteType(...routeTypes);
}
exports.addRouteType = addRouteType;
/**
 * @param {Manifest} manifest
 * @param {...any} routes
 */
function addRoute(manifest, ...routes) {
    manifest.addRoute(...routes);
}
exports.addRoute = addRoute;
/**
 * @param {Manifest} manifest
 * @param {...any} contents
 */
function addContent(manifest, ...contents) {
    manifest.addContent(...contents);
}
exports.addContent = addContent;
/**
 * @param {Manifest} manifest
 * @param {...any} entries
 */
function addDictionary(manifest, ...entries) {
    manifest.addDictionary(...entries);
}
exports.addDictionary = addDictionary;
