// @sitecore-jss/sitecore-jss/dist/cjs/index.js
"use strict";
// NOTE: all imports are now named as to not make breaking changes
// and to keep react-native working with cjs modules.
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.constants = exports.NativeDataFetcher = exports.AxiosDataFetcher = exports.GraphQLRequestClient = exports.fetchData = exports.enableDebug = exports.debug = void 0;
const constants = __importStar(require("./constants"));
exports.constants = constants;
var debug_1 = require("./debug");
Object.defineProperty(exports, "debug", { enumerable: true, get: function () { return __importDefault(debug_1).default; } });
Object.defineProperty(exports, "enableDebug", { enumerable: true, get: function () { return debug_1.enableDebug; } });
var data_fetcher_1 = require("./data-fetcher");
Object.defineProperty(exports, "fetchData", { enumerable: true, get: function () { return data_fetcher_1.fetchData; } });
var graphql_request_client_1 = require("./graphql-request-client");
Object.defineProperty(exports, "GraphQLRequestClient", { enumerable: true, get: function () { return graphql_request_client_1.GraphQLRequestClient; } });
var axios_fetcher_1 = require("./axios-fetcher");
Object.defineProperty(exports, "AxiosDataFetcher", { enumerable: true, get: function () { return axios_fetcher_1.AxiosDataFetcher; } });
var native_fetcher_1 = require("./native-fetcher");
Object.defineProperty(exports, "NativeDataFetcher", { enumerable: true, get: function () { return native_fetcher_1.NativeDataFetcher; } });
