// @sitecore-jss/sitecore-jss/dist/cjs/utils/index.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.handleEditorAnchors = exports.resetEditorChromes = exports.isEditorActive = exports.HorizonEditor = exports.ExperienceEditor = exports.isTimeoutError = exports.isAbsoluteUrl = exports.resolveUrl = exports.isServer = void 0;
var is_server_1 = require("./is-server");
Object.defineProperty(exports, "isServer", { enumerable: true, get: function () { return __importDefault(is_server_1).default; } });
var utils_1 = require("./utils");
Object.defineProperty(exports, "resolveUrl", { enumerable: true, get: function () { return utils_1.resolveUrl; } });
Object.defineProperty(exports, "isAbsoluteUrl", { enumerable: true, get: function () { return utils_1.isAbsoluteUrl; } });
Object.defineProperty(exports, "isTimeoutError", { enumerable: true, get: function () { return utils_1.isTimeoutError; } });
var editing_1 = require("./editing");
Object.defineProperty(exports, "ExperienceEditor", { enumerable: true, get: function () { return editing_1.ExperienceEditor; } });
Object.defineProperty(exports, "HorizonEditor", { enumerable: true, get: function () { return editing_1.HorizonEditor; } });
Object.defineProperty(exports, "isEditorActive", { enumerable: true, get: function () { return editing_1.isEditorActive; } });
Object.defineProperty(exports, "resetEditorChromes", { enumerable: true, get: function () { return editing_1.resetEditorChromes; } });
Object.defineProperty(exports, "handleEditorAnchors", { enumerable: true, get: function () { return editing_1.handleEditorAnchors; } });
