// @sitecore-jss/sitecore-jss-cli/dist/cjs/scripts/deploy.config.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.handler = exports.builder = exports.describe = exports.command = void 0;
/* eslint-disable prettier/prettier */
const sitecore_jss_dev_tools_1 = require("@sitecore-jss/sitecore-jss-dev-tools");
const path_1 = __importDefault(require("path"));
const resolve_package_1 = __importDefault(require("../resolve-package"));
exports.command = 'config';
exports.describe = 'Deploys the app\'s Sitecore configuration file(s) to the Sitecore server. `jss deploy config --help` for options.';
exports.builder = {
    source: {
        requiresArg: false,
        type: 'string',
        describe: 'The source path of the config patches to deploy.',
        default: './sitecore/config',
    },
    destination: {
        requiresArg: false,
        type: 'string',
        describe: 'Destination path to deploy to. Defaults to the \'instancePath\' set in scjssconfig.json, combined with the \'sitecoreConfigPath\' setting from package.json.',
    },
    config: {
        requiresArg: false,
        type: 'string',
        describe: 'Path to scjssconfig file.',
        default: './scjssconfig.json',
    },
};
/**
 * @param {any} argv
 */
// eslint-disable-next-line @typescript-eslint/no-explicit-any
function handler(argv) {
    return __awaiter(this, void 0, void 0, function* () {
        sitecore_jss_dev_tools_1.verifySetup();
        const options = {
            destinationPath: argv.destination,
            sourcePath: argv.source,
            clean: false,
        };
        if (!options.destinationPath) {
            const packageJson = yield resolve_package_1.default();
            const scJssConfig = yield sitecore_jss_dev_tools_1.resolveScJssConfig({ configPath: argv.config });
            options.destinationPath = path_1.default.join(scJssConfig.sitecore.instancePath, packageJson.config.sitecoreConfigPath);
        }
        sitecore_jss_dev_tools_1.deploy(options);
    });
}
exports.handler = handler;
