// @sitecore-jss/sitecore-jss-cli/dist/cjs/scripts/deploy.package.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.handler = exports.builder = exports.describe = exports.command = void 0;
const package_1 = require("./package");
exports.command = 'package';
exports.describe = null;
exports.builder = Object.assign({ deployUrl: {
        requiresArg: false,
        type: 'string',
    }, skipPackage: {
        requiresArgs: false,
        type: 'boolean',
        default: false,
    } }, package_1.builder);
/**
 * Deploy package handler
 */
function handler() {
    return __awaiter(this, void 0, void 0, function* () {
        console.error('jss deploy package is no longer used. Use jss deploy app instead.');
        process.exit(1);
    });
}
exports.handler = handler;
