// @sitecore-jss/sitecore-jss-cli/dist/cjs/scripts/package.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.handler = exports.builder = exports.describe = exports.command = void 0;
const sitecore_jss_dev_tools_1 = require("@sitecore-jss/sitecore-jss-dev-tools");
const resolve_package_1 = __importDefault(require("../resolve-package"));
const manifest_1 = require("./manifest");
exports.command = 'package';
exports.describe = 'Generates a JSS manifest package file that can be deployed to Sitecore. Nothing is deployed. See also jss deploy package, which takes the package and deploys it to Sitecore. `jss package --help` for options.';
exports.builder = Object.assign({ packageOutputPath: {
        requiresArgs: false,
        type: 'string',
        describe: 'The location the JSS manifest package will be written to.',
        default: './sitecore/package',
    }, skipManifest: {
        requiresArgs: false,
        type: 'boolean',
        describe: 'If true, skips manifest generation. This can be used to consume existing output from jss manifest (via the manifestOutputPath parameter) without rebuilding it.',
        default: false,
    } }, manifest_1.builder);
/**
 * @param {any} argv
 */
// eslint-disable-next-line @typescript-eslint/no-explicit-any
function handler(argv) {
    return __awaiter(this, void 0, void 0, function* () {
        const packageJson = yield resolve_package_1.default();
        if (!argv.appName) {
            argv.appName = packageJson.config.appName;
        }
        if (!argv.appName) {
            throw new Error('App Name was not defined as a parameter or in the package.json config');
        }
        let continuation = Promise.resolve();
        // if we need items, generate the manifest
        if (!argv.noItems && !argv.skipManifest) {
            continuation = manifest_1.handler(argv);
        }
        return continuation.then(() => {
            const generateArgs = {
                appName: argv.appName,
                manifestPath: argv.manifestOutputPath,
                manifestFileName: 'sitecore-import.json',
                outputPath: argv.packageOutputPath,
            };
            console.log(`JSS is manifest packaging ${argv.appName} to ${argv.packageOutputPath}...`);
            return sitecore_jss_dev_tools_1.packageGenerate(generateArgs);
        });
    });
}
exports.handler = handler;
