// @sitecore-jss/sitecore-jss-dev-tools/dist/cjs/manifest/generator/pipelines/generateRouteItem/pipeline.config.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.config = void 0;
const index_1 = require("./../../../../index");
const path = __importStar(require("path"));
const dynamicPlaceholders_1 = require("../../dynamicPlaceholders");
const processRenderings_1 = __importDefault(require("./processRenderings"));
// __dirname returns the directory of this file/module, so it has to be called here
const resolveModulePath = (modulePath) => path.resolve(__dirname, modulePath);
const defaultDatasourceNamer = ({ item, placeholder, rendering, index, }) => {
    const placeholderKeys = placeholder.phKey.split('/');
    let finalPlaceholderKey;
    if (placeholderKeys.length === 1) {
        // single named placeholder e.g. /main
        finalPlaceholderKey = placeholderKeys[0];
    }
    else {
        // dynamic placeholder. We want to extract the final DP key as it should be unique among all DPs.
        // e.g. '/main/footer-bottom-{B4C653AA-FFBC-5821-A8A8-386E791553FE}-0' => 'footer-bottom-B4C653AA-FFBC-5821-A8A8-386E791553FE-0'
        finalPlaceholderKey = placeholderKeys[placeholderKeys.length - 1];
        // get rid of braces
        finalPlaceholderKey = finalPlaceholderKey.replace(/({|})/g, '');
    }
    return `${item.name}-${finalPlaceholderKey}-${rendering.componentName}-${index + 1}`;
};
const defaultDatasourceDisplayNamer = ({ rendering, index, }) => `${rendering.componentName}-${index + 1}`;
const config = (pipelines) => {
    const pipeline = index_1.pipelineFactory.create('generateRouteItem');
    pipeline.addProcessor({
        name: 'generateItem',
        modulePath: resolveModulePath('./generateItem.js'),
    });
    pipeline.addProcessor({
        name: 'processPlaceholders',
        modulePath: resolveModulePath('./processPlaceholders.js'),
    });
    pipeline.addProcessor({
        name: 'processRenderings',
        process: processRenderings_1.default,
        args: {
            dynamicPlaceholderKeyGenerator: dynamicPlaceholders_1.getDynamicPlaceholderKey,
            placeholder: {
                phKey: '/',
            },
            rendering: {},
            datasourceNamer: defaultDatasourceNamer,
            datasourceDisplayNamer: defaultDatasourceDisplayNamer,
        },
    });
    pipeline.addProcessor({
        name: 'processChildRoutes',
        modulePath: resolveModulePath('./processChildRoutes.js'),
    });
    pipelines.addPipeline(pipeline);
};
exports.config = config;
