// @sitecore-jss/sitecore-jss-dev-tools/dist/esm/manifest/generator/manifest.types.js
export var CommonFieldTypes;
(function (CommonFieldTypes) {
    CommonFieldTypes["SingleLineText"] = "Single-Line Text";
    CommonFieldTypes["MultiLineText"] = "Multi-Line Text";
    CommonFieldTypes["RichText"] = "Rich Text";
    CommonFieldTypes["ContentList"] = "Treelist";
    CommonFieldTypes["ItemLink"] = "Droptree";
    CommonFieldTypes["GeneralLink"] = "General Link";
    CommonFieldTypes["Image"] = "Image";
    CommonFieldTypes["File"] = "File";
    CommonFieldTypes["Number"] = "Number";
    CommonFieldTypes["Checkbox"] = "Checkbox";
    CommonFieldTypes["Date"] = "Date";
    CommonFieldTypes["DateTime"] = "Datetime";
})(CommonFieldTypes || (CommonFieldTypes = {}));
export var FieldStorage;
(function (FieldStorage) {
    FieldStorage["Versioned"] = "versioned";
    FieldStorage["Shared"] = "shared";
    FieldStorage["Unversioned"] = "unversioned";
})(FieldStorage || (FieldStorage = {}));
/**
 * @param {ItemDefinition | ItemReference} obj
 */
export function isItemDefinition(obj) {
    return obj.name !== undefined;
}
