// @sitecore-jss/sitecore-jss-dev-tools/dist/esm/manifest/generator/manifest.babel-shim.js
// this is a shim to allow non-TypeScript consumers to take advantage of typings data when they cannot specify the
// type of the manifest var in a manifest source file, e.g.
// instead of export default (manifest: Manifest) => { ... }, you can use
// import { addComponent } from 'thislib';
// export default (manifest) => { addComponent(manifest, { data }); }
// DEPRECATED: favor using typing JSDoc instead, e.g. @param {Manifest} manifest Manifest instance to add components to
/**
 * @param {Manifest} manifest
 * @param {...any} components
 */
export function addComponent(manifest, ...components) {
    manifest.addComponent(...components);
}
/**
 * @param {Manifest} manifest
 * @param {...any} templates
 */
export function addTemplate(manifest, ...templates) {
    manifest.addTemplate(...templates);
}
/**
 * @param {Manifest} manifest
 * @param {...any} placeholders
 */
export function addPlaceholder(manifest, ...placeholders) {
    manifest.addPlaceholder(...placeholders);
}
/**
 * @param {Manifest} manifest
 * @param {...any} routeTypes
 */
export function addRouteType(manifest, ...routeTypes) {
    manifest.addRouteType(...routeTypes);
}
/**
 * @param {Manifest} manifest
 * @param {...any} routes
 */
export function addRoute(manifest, ...routes) {
    manifest.addRoute(...routes);
}
/**
 * @param {Manifest} manifest
 * @param {...any} contents
 */
export function addContent(manifest, ...contents) {
    manifest.addContent(...contents);
}
/**
 * @param {Manifest} manifest
 * @param {...any} entries
 */
export function addDictionary(manifest, ...entries) {
    manifest.addDictionary(...entries);
}
